src/core/transpile.jl

```diff
@@ -479,3 +484,126 @@ function transpile(::CastToPhaseShiftAndHalfRotationX, circuit::QuantumCircuit):
+final operator acts on adjacent qubits. The result of the input 
+and output circuit on any arbitrary state Ket is unchanged 
+(up to a global phase).
+
+# Examples
+```jldoctest
+julia> transpiler=Snowflake.PlaceOperationsOnLine();
+
+julia> circuit = QuantumCircuit(qubit_count = 6, gates=[toffoli(4,6,1)])
+Quantum Circuit Object:
+   qubit_count: 6 
+q[1]:──X──
+       |  
+q[2]:──|──
+       |  
+q[3]:──|──
+       |  
+q[4]:──*──
+       |  
+q[5]:──|──
+       |  
+q[6]:──*──
+          
+
+
+
+
+julia> transpiled_circuit=transpile(transpiler,circuit)
+Quantum Circuit Object:
+   qubit_count: 6 
+q[1]:───────────────────────────X───────────────────────────
+                                |                           
+q[2]:───────☒───────────────────*───────────────────☒───────
+            |                   |                   |       
+q[3]:──☒────☒──────────────☒────*────☒──────────────☒────☒──
+       |                   |         |                   |  
+q[4]:──☒──────────────☒────☒─────────☒────☒──────────────☒──
+                      |                   |                 
+q[5]:────────────☒────☒───────────────────☒────☒────────────
+                 |                             |            
+q[6]:────────────☒─────────────────────────────☒────────────
+                                                            
+
+
+
+julia> compare_circuits(circuit,transpiled_circuit)
+true
+
+```
+"""
+function transpile(::PlaceOperationsOnLine, circuit::QuantumCircuit)::QuantumCircuit
+
+    gates=get_circuit_gates(circuit)
+    
+    qubit_count=get_num_qubits(circuit)
+    output_circuit=QuantumCircuit(qubit_count=qubit_count)
+
+    for gate in gates
+
+        connected_qubits=sort(get_connected_qubits(gate))
+
+        if length(connected_qubits)>1
+
+            targets=get_targets(gate)
+    
+            # in controlled gates, length(target_indices)<length(connected_qubits)
+            targets_indices=findall(x->x in targets,connected_qubits)
+    
+            # left untouched
+            min_target=connected_qubits[1] 
+
+            compressed_targets=[min_target+i for i in 0:(length(connected_qubits)-1)]
+
+            previous_min_target=min_target
+
+            # for controlled gates, target qubit is last argument
+            connected_qubits_output=vcat(
+                [p for (i,p) in enumerate(compressed_targets) if !(i in targets_indices)],
+                compressed_targets[targets_indices] )
+                
+            gates_block=[typeof(gate)(connected_qubits_output...,)]
+
+            @assert get_targets(gates_block[1])==compressed_targets[targets_indices] (
+                "Failed to construct gate: $(typeof((gates_block[1])))")
+
+            compressed_targets=reverse(compressed_targets)
+
+            for (i,next_target) in enumerate(reverse(connected_qubits[2:end]))
+                nearest_target=compressed_targets[i]
+
+                while !isequal(next_target,nearest_target)
```

Comment:
This function is getting complex. We could factor out a qubit remap function which takes a new qubit order and adds swaps to get there. For example

`multi_swap_qubits!(circuit::QuantumCircuit, qubit_mapping::AbstractVector{Int})`

for example with
`circuit = circuit::QuantumCircuit(qubit_count=3, gates=[...])`

when
`multi_swap_qubits(circuit, [3,1,2])`

will add swaps to `circuit` so that qubit 3 is now at position 1, qubit 1 is now at 2, and qubit 2 is not at position 3.

Does it make sense what I am trying to describe?